test/devapps/username_password_sample.go

```diff
@@ -4,48 +4,50 @@
 package main
 
 import (
+	"context"
 	"fmt"
 
+	"github.com/AzureAD/microsoft-authentication-library-for-go/internal/msalbase"
 	"github.com/AzureAD/microsoft-authentication-library-for-go/msal"
 	log "github.com/sirupsen/logrus"
 )
 
-func tryUsernamePasswordFlow(publicClientApp *msal.PublicClientApplication) {
-	userNameParams := msal.CreateAcquireTokenUsernamePasswordParameters(config.Scopes, config.Username, config.Password)
-	result, err := publicClientApp.AcquireTokenByUsernamePassword(userNameParams)
-	if err != nil {
-		log.Fatal(err)
-	}
-	accessToken := result.GetAccessToken()
-	log.Info("Access token is: " + accessToken)
-}
-
 func acquireByUsernamePasswordPublic() {
 	config := createConfig("config.json")
-	// Creating the Public Client Application
-	publicClientApp, err := msal.CreatePublicClientApplication(config.ClientID, config.Authority)
-	if err != nil {
-		log.Fatal(err)
-	}
-	publicClientApp.SetCacheAccessor(cacheAccessor)
-	var userAccount msal.AccountProvider
-	accounts := publicClientApp.GetAccounts()
+	// create a PublicClientApplication with a  custom cache accessor
+	options := msal.DefaultPublicClientApplicationOptions()
+	options.Accessor = cacheAccessor
+	publicClientApp := msal.NewPublicClientApplication(config.ClientID, config.Authority, &options)
+
+	// look in the cache to see if the account to use has been cached
+	var userAccount *msalbase.Account
+	accounts := publicClientApp.Accounts()
 	for _, account := range accounts {
-		if account.GetUsername() == config.Username {
+		if account.PreferredUsername == config.Username {
 			userAccount = account
 		}
 	}
-	if userAccount == nil {
-		log.Info("No valid account found")
-		tryUsernamePasswordFlow(publicClientApp)
-	} else {
-		silentParams := msal.CreateAcquireTokenSilentParametersWithAccount(config.Scopes, userAccount)
-		result, err := publicClientApp.AcquireTokenSilent(silentParams)
+	var result *msalbase.AuthenticationResult
+	var err error
+	if userAccount != nil {
+		// found a cached account, now see if an applicable token has been cached
+		// NOTE: this API conflates error states, i.e. err is non-nil if an applicable token isn't
```

Comment:
Any suggestions on this?

Reply:
I've been thinking about this one a bit.  It's somewhat similar to calling an [io.Reader](https://golang.org/pkg/io/#Reader) returning `io.EOF` at the end of a stream.

I'm not familiar enough with how customers use this functionality.  Does it seem valuable to return a distinct error, e.g. `msal.ErrNoToken` (something like that)?  Are there use-cases where distinguishing this specific case is valuable?

Reply:
MSAL errors codes are mostly well known. e.g. it may be a MsalClientException, an MsalServiceException. And there are specific error codes that lets a developer write the code in a specific way. we should do the same here. Java, Python, Node and .NET show cases this

---
In reply to: [515436829](https://github.com/AzureAD/microsoft-authentication-library-for-go/pull/68#discussion_r515436829) [](ancestors = 515436829)

Reply:
I took a cursory look at the C# implementation and it doesn't appear to throw an exception if the account isn't found (it's not doc'ed that way, and I dug a bit into the implementation and didn't see anyplace where it obviously throws).

Reply:
GetAccounts doesn't throw. My comment was arround the API in general (sorry for not being clear about that). My main comment on errors was that we should take that as a separate work item (already tracked elsewhere). We should simply follow the std. msal patterns for user flows and that will lead to specific expected errors that guide the developer and the user experience.